### PR TITLE
[FLINK-10866][Runtime] 1. Explicitly enable qs server and proxy. 2. QS

### PR DESCRIPTION
**Pull request by eaglewatcherwb**
startup failure not block TM startup

Change-Id: Id79c8bf97002a387a80be563a43dce3210143dc2

## What is the purpose of the change

1. Explicitly enable qs server and proxy.
2. QS startup failure not block TM startup.

## Brief change log

1. add `queryable-state.enable` configuration to indicate whether start QS server and proxy, the default value of which is false.
2. QS server and proxy startup failures does not cause TM startup failure any more.

## Verifying this change

This change added tests and can be verified as follows:
1. With default configuration, QS server and proxy does not start.
2. Set query.enable to true, QS server and proxy start.
3. QS server or proxy does not cause TM failure when got startup failures.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): no
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`:  no
  - The serializers: no
  - The runtime per-record code paths (performance sensitive):  no
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Yarn/Mesos, ZooKeeper: no
  - The S3 file system connector: no 

## Documentation

  - Does this pull request introduce a new feature? no
  - If yes, how is the feature documented? (not applicable / docs / JavaDocs / not documented)

### CODE REVIEW

File: flink-core/src/main/java/org/apache/flink/configuration/QueryableStateOptions.java [hunk 1]
```diff
@@ -106,6 +106,12 @@
 			.defaultValue(0)
 			.withDescription("Number of query Threads for queryable state server. Uses the number of slots if set to 0.");
 
+	/** Enable Queryable state server and proxy. */
+	public static final ConfigOption<Boolean> ENABLE_QUERYABLE_STATE_PROXY_SERVER =
+		key("query.enable")
```
Comment (klion26) [inline flink-core/src/main/java/org/apache/flink/configuration/QueryableStateOptions.java]: how about changing the key to `query.server.xxx` as above, and the description to "Option whether ..." as `CheckpointingOptions.ASYNC_SNAPSHOTS`

File: flink-runtime/src/main/java/org/apache/flink/runtime/taskexecutor/TaskManagerServicesConfiguration.java [hunk 2]
```diff
@@ -107,7 +107,7 @@ public TaskManagerServicesConfiguration(
 		this.localRecoveryStateRootDirectories = checkNotNull(localRecoveryStateRootDirectories);
 		this.localRecoveryEnabled = checkNotNull(localRecoveryEnabled);
 		this.networkConfig = checkNotNull(networkConfig);
-		this.queryableStateConfig = checkNotNull(queryableStateConfig);
+		this.queryableStateConfig = queryableStateConfig;
```
Comment (klion26) [inline flink-runtime/src/main/java/org/apache/flink/runtime/taskexecutor/TaskManagerServicesConfiguration.java]: Should we add `@Nullable` to the parameter `queryableStateConfig`?

**Comment by buptljy** [comment]
@eaglewatcherwb Thanks for this PR!
Agree to the "query-server.enable" configuration, but I'm not sure that we should prevent all the failures which are caused by this, because the queryable state may be a main feature for some business. 

I wonder that if we can enable this server by checking the registeredKvStates so that we can know if the uses are really using the feature.

File: flink-core/src/main/java/org/apache/flink/configuration/QueryableStateOptions.java (hunk 1 above)
Comment (eaglewatcherwb) [reply flink-core/src/main/java/org/apache/flink/configuration/QueryableStateOptions.java]: ENABLE_QUERYABLE_STATE_PROXY_SERVER is used to configure both QS proxy and server, so I think query.enable is more meaningful.
Make description more readable is a good idea, I will change it.

File: flink-runtime/src/main/java/org/apache/flink/runtime/taskexecutor/TaskManagerServicesConfiguration.java (hunk 2 above)
Comment (eaglewatcherwb) [reply flink-runtime/src/main/java/org/apache/flink/runtime/taskexecutor/TaskManagerServicesConfiguration.java]: Good idea, I will add it.

**Comment by eaglewatcherwb** [comment]
> @eaglewatcherwb Thanks for this PR!
> Agree to the "query-server.enable" configuration, but I'm not sure that we should prevent all the failures which are caused by this, because the queryable state may be a main feature for some business.
> 
> I wonder that if we can enable this server by checking the registeredKvStates so that we can know if the uses are really using the feature.

Thanks for the comments.
**checking the registeredKvStates** do you mean **registerQueryableState** in TaskExecutor ?

**Comment by eaglewatcherwb** [comment]
> > > @eaglewatcherwb Thanks for this PR!
> > > Agree to the "query-server.enable" configuration, but I'm not sure that we should prevent all the failures which are caused by this, because the queryable state may be a main feature for some business.
> > > I wonder that if we can enable this server by checking the registeredKvStates so that we can know if the uses are really using the feature.
> > 
> > 
> > @buptljy Thanks for the comments.
> > **checking the registeredKvStates** do you mean **registerQueryableState** in TaskExecutor ?
> 
> The queryable state will be registered in KvStateRegistry when it is published, which enables you to know if users have queryable state, with kvStateRegistry in NetworkEnvironment.

kvStateRegistry in NetworkEnvironment is always associated with JobId and the queryable state is registered when a task runs on TM, however, QS server is a service of TM, which is independent of job or task and a common service of TM,  and is initialized before a task runs on TM. Thus, I think it may be improper to enable QS service by checking the registeredKvStates, and should by user explicit configuration.

File: flink-core/src/main/java/org/apache/flink/configuration/QueryableStateOptions.java [hunk 3]
```diff
@@ -109,7 +109,7 @@
 	/** Option whether the queryable state proxy and server should be enabled where possible and configurable.
 	 *
 	 * <p>Queryable state proxy and server are still more experimental features, hence disabled unless they are enable
-	 * in user configuration. */
+	 * in user configuration.  */
```
Comment (klion26) [inline flink-core/src/main/java/org/apache/flink/configuration/QueryableStateOptions.java]: Maybe we can remove the added blank character.

Comment (eaglewatcherwb) [reply flink-core/src/main/java/org/apache/flink/configuration/QueryableStateOptions.java]: > Maybe we can remove the added blank character.

Very careful code review :)
I will remove it.

File: flink-runtime/src/test/java/org/apache/flink/runtime/testutils/MiniClusterResource.java [hunk 4]
```diff
@@ -113,6 +115,7 @@ public void after() {
 
 	private void startMiniCluster() throws Exception {
 		final Configuration configuration = new Configuration(miniClusterResourceConfiguration.getConfiguration());
+		configuration.setBoolean(ENABLE_QUERYABLE_STATE_PROXY_SERVER, true);
```
[the remark this answers is not in the document]
Comment (eaglewatcherwb) [reply flink-runtime/src/test/java/org/apache/flink/runtime/testutils/MiniClusterResource.java]: ok, queryable state should be activated only in queryable related cases.
Configuration using **queryable-state** is a good idea, I will use it.

[the remark this answers is not in the document]
Comment (klion26) [reply flink-runtime/src/test/java/org/apache/flink/runtime/testutils/MiniClusterResource.java]: @eaglewatcherwb Should we enable the queryable-state in somewhere? Or use Parameterized to test both enabled and disabled scenario

[the remark this answers is not in the document]
Comment (eaglewatcherwb) [reply flink-runtime/src/test/java/org/apache/flink/runtime/testutils/MiniClusterResource.java]: > @eaglewatcherwb Should we enable the queryable-state in somewhere? Or use Parameterized to test both enabled and disabled scenario

Yes, queryable-state is only enabled in related cases now.

**Comment by aljoscha** [comment]
This looks good now! I'll merge once Travis is green.

@eaglewatcherwb Do you want to open a follow-up Jira issue to rename the other options from `query.x` to `queryable-state.x`. We can keep the old options around with the `deprecatedKey()` method for now.

**Comment by aljoscha** [comment]
@eaglewatcherwb Are both `eaglewatcherwb` and `yanyu.wb` your nicknames? I want to squash the git commits into one commit but which username should be on the final commit?

**Comment by eaglewatcherwb** [comment]
> @eaglewatcherwb Are both `eaglewatcherwb` and `yanyu.wb` your nicknames? I want to squash the git commits into one commit but which username should be on the final commit?

Yes, both are my nicknames. eaglewatcherwb is OK

**Comment by eaglewatcherwb** [comment]
> This looks good now! I'll merge once Travis is green.
> 
> @eaglewatcherwb Do you want to open a follow-up Jira issue to rename the other options from `query.x` to `queryable-state.x`. We can keep the old options around with the `deprecatedKey()` method for now.

Ok, I am glad to do this. The [follow-up jira](https://issues.apache.org/jira/browse/FLINK-11295) is open.

**Comment by aljoscha** [comment]
Merged, thanks again @eaglewatcherwb! 😄

**Pull request closed without merging**